doc/design/model_zoo.md

```diff
@@ -0,0 +1,104 @@
+   1. **Training hyperparameters** appear in WITH, as model constructors do. They affect the training process.
+   1. **Data conversion hyperparameters** appear in COLUMN and optionally LABEL. For example, the scale range and bucket size in the first example.
+
+1. A **trained model** consists of all the above concepts and the estimated model parameters.
+
+## Representation
+
+### Versioning and Releasing
+
+A key to this design is the representation of the above concepts.  It is not a straightforward solution. For example, in prior work and previous discussions, engineers proposed to represent each model definition by a Python source code file hosted on an FTP service and downloaded when some users file a model training statement. This intuitive proposal could lead to inconsistencies. Suppose that an analyst trained a model using the definition in `my_dnn_regressor.py` and got `my_first_model`; soon after that, a data scientist changed `my_dnn_regressor.py` and the analyst re-trained the model into `my_second_model` with slightly modified hyperparameter settings. The analyst might expect that both models share the same definition; however, they don't, and worse than that, there is no mechanism to remind the change of the model definition to the analyst.
+
+Such design flaw roots from the ignoring of the fact that model definitions are code that has versions. Another important aspect is that we often build the source code into some release-ready form.  Once we noticed these facts, we can use version management tools and Git and release engineering tools like Docker.  Here follows our proposal.
+
+1. A collection of model definitions is a Git repository of source files. 
+1. To describe dependencies, we require a Dockerfile at the root directory of the repository.
+1. To release a repository, we checkout the specific version and run `docker run` with the Dockerfile.
+1. Each release has a globally unique ID (GUID) represented by the hash ID of the Docker image.
+1. The Docker image must support a protocol of command-line arguments so that SQLFlow can pass in required information like hyperparameters from the `docker run` command.
+
+### Argument Protocol
+
+For a SELECT program using the SQLFlow syntax extension, the SQLFlow server converts it into a *submitter* program, usually in Python.
+
+Consider the training statement at the beginning of this document. Suppose that the Python class `MyDNNRegressor` is in source file `my_dnn_regressor.py` in the repository https://github.com/a_data_scientist/regressors, which builds into Docker image `a_data_scientist/regressors`.  Let us also assume that the submitter program is `/var/sqlflow/submitters/an_analyst/my_first_model.py`.  SQLFlow can run the submitter using the Docker image as follows:
+
+```bash
+docker run --rm -it \
+  -v /var/sqlflow/submitters:/submitters \
+  a_data_scientist/regressors \
+    python /submitters/an_analyst/my_first_model.py
```

Comment:
For distributed cases, the command should be calling ElasticDL. Also how to put generated `my_first_model.py` to the runtime containers.

Reply:
In my mind, in order to submit an ElasticDL training job, we need the following actions.

1. The SQLFlow server invokes codegen_elasticdl.go to generates the submitter program `my_first_model.py`.
1. `my_first_model.py` calls the ElasticDL client API to submit a job.
1. The SQLFlow server then runs `my_first_model.py` in a Docker container (if SQLFlow service is not running in a Kubernetes cluster), or a Pod, which has ElasticDL client library installed.
1. As this container needs to contain ElasticDL and the model definition, it must be derived from `a_data_scientist/regressor`, like from the following Dockerfile:

   ```dockerfile
   FROM a_data_scientist/regressors
   RUN pip install -y elasticdl_api
   ```

Please correct me or improve the above idea.

Reply:
I see, that's clear enough. One question, maybe we can just build all model zoo Docker images with ElasticDL API ( model Dockerfile should depend on our base image like `sqlflow/sqlflow_model_base`), then the submitted job can directly use the model's Docker image, or else it may require to build an image and push to the registry before submitting the job.

Reply:
I see. It is like requiring all model definition repositories have their Dockefile's all derived from a common base image. I think it is a viable solution.

Reply:
I also had the same question as @typhoonzero while I am reading the design. Please add "the following actions" to the design doc.

Reply:
> In my mind, in order to submit an ElasticDL training job, we need the following actions.
> 
> 1. The SQLFlow server invokes codegen_elasticdl.go to generates the submitter program `my_first_model.py`.
> 2. `my_first_model.py` calls the ElasticDL client API to submit a job.
> 3. The SQLFlow server then runs `my_first_model.py` in a Docker container (if SQLFlow service is not running in a Kubernetes cluster), or a Pod, which has ElasticDL client library installed.
> 4. As this container needs to contain ElasticDL and the model definition, it must be derived from `a_data_scientist/regressor`, like from the following Dockerfile:
>    ```dockerfile
>    FROM a_data_scientist/regressors
>    RUN pip install -y elasticdl_api
>    ```
> 
> Please correct me or improve the above idea.

Are you talking about submitting a model to the zoo, or submitting a training job? A little bit lost in the discussion.